rasa_nlu/server.py

```diff
@@ -48,74 +46,110 @@ def requires_auth(f):
-                response = current_app.data_router.parse(data)
-                return jsonify(response)
-            except InvalidModelError as e:
-                return jsonify({"error": "{}".format(e)}), 404
-
-    @rasa_nlu_app.route("/version", methods=['GET'])
-    @requires_auth
-    def version():
-        return jsonify({'version': __version__})
+            def errback(f):
+                f.trap(InvalidModelError)
 
-    @rasa_nlu_app.route("/config", methods=['GET'])
+                request.setResponseCode(404)
+                request.setHeader('Content-Type', 'application/json')
+                return json.dumps({"error": "{}".format(f.getErrorMessage())})
+
+            data = self.data_router.extract(request_params)
+            response = threads.deferToThread(self.data_router.parse, data)
+            request.setHeader('Content-Type', 'application/json')
+            response.addCallback(json.dumps)
+            response.addErrback(errback)
+            return response
+
+    @app.route("/version", methods=['GET'])
     @requires_auth
-    def rasaconfig():
-        return jsonify(config.as_dict())
+    def version(self, request):
+        """Returns the Rasa server's version"""
 
-    @rasa_nlu_app.route("/status", methods=['GET'])
+        request.setHeader('Content-Type', 'application/json')
+        return json.dumps({'version': __version__})
+
+    @app.route("/config", methods=['GET'])
     @requires_auth
-    def status():
-        return jsonify(current_app.data_router.get_status())
+    def rasaconfig(self, request):
+        """Returns the in-memory configuration of the Rasa server"""
 
-    @rasa_nlu_app.route("/", methods=['GET'])
-    def hello():
-        return "hello from Rasa NLU: " + __version__
+        request.setHeader('Content-Type', 'application/json')
+        return json.dumps(self.config.as_dict())
 
-    @rasa_nlu_app.route("/train", methods=['POST'])
+    @app.route("/status", methods=['GET'])
     @requires_auth
-    def train():
-        data_string = request.get_data(as_text=True)
-        current_app.data_router.start_train_process(data_string, request.args)
-        return jsonify(info="training started.", training_process_ids=current_app.data_router.train_proc_ids())
+    def status(self, request):
+        request.setHeader('Content-Type', 'application/json')
+        return json.dumps(self.data_router.get_status())
+
+    @app.route("/train", methods=['POST'])
+    @requires_auth
+    def train(self, request):
+        def errback(f):
+            f.trap(ValueError)
+            logger.debug("error: {}".format(f.getErrorMessage()))
+
+        data_string = request.content.read().decode('utf-8', 'strict')
 
-    logging.basicConfig(filename=config['log_file'], level=config['log_level'])
-    logging.captureWarnings(True)
-    logger.info("Configuration: " + config.view())
+        test = self.data_router.start_train_process(data_string,
```

Comment:
Of course. I called it test because it just output the raw error message to the user if the server encounters a failure. Maybe this could be improved.